src/hypothesis/strategies.py

```diff
@@ -334,17 +334,35 @@ def floats(
                     allow_nan
                 ))
 
-    min_value = try_convert(float, min_value, 'min_value')
-    max_value = try_convert(float, max_value, 'max_value')
-
     check_valid_bound(min_value, 'min_value')
     check_valid_bound(max_value, 'max_value')
+
+    min_arg, max_arg = min_value, max_value
+    if min_value is not None:
+        min_value = float(min_value)
+    if max_value is not None:
+        max_value = float(max_value)
+
     check_valid_interval(min_value, max_value, 'min_value', 'max_value')
     if min_value == float(u'-inf'):
         min_value = None
     if max_value == float(u'inf'):
         max_value = None
 
+    if min_value is not None and min_value < min_arg:
+        min_value = next_up(min_value)
+        assert min_value > min_arg
+    if max_value is not None and max_value > max_arg:
+        max_value = next_down(max_value)
+        assert max_value < max_arg
+    if None not in (min_value, max_value) and min_value > max_value:
```

Comment:
Would we be okay with making this an error? The previous behavior (giving any examples) was wrong, and I think we said we'll (somethimes/often/?) be okay with turning wrong behavior into an error?

The case with the unicode categories was a bit different because we had the thing we wanted to make an error right in our documentation, but to me this one seems fair to turn into an error right away.

Reply:
Sounds good to me, though I've left it as a separate commit in case @DRMacIver wants it reverted.